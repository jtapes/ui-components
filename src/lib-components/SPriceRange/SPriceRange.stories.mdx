import {action} from '@storybook/addon-actions'
import {config, withDesign} from 'storybook-addon-designs'
import {withTests} from '@storybook/addon-jest';
import {
    Meta,
    Story,
    ArgsTable,
    Description,
    Source,
    Canvas,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SPriceRange from '@/lib-components/SPriceRange/SPriceRange.vue';

<Meta
    title="DESIGN SYSTEM/Atom/PriceRange"
    component={SPriceRange}
    decorators={[withDesign, withTests({results})]}
    argTypes={{
        value: {
            control: 'array',
            defaultValue: [10, 30],
            description: 'Диапозон цен',
            type: 'Array',
        },
        max: {
            control: 'number',
            defaultValue: 100,
            description: 'Максимальная цена',
            type: 'Number'
        },
        min: {
            control: 'number',
            defaultValue: 0,
            description: 'Минимальная цена',
            type: 'Number'
        },
    }}
    parameters={{
        design: config({
            type: 'figma',
            url:
                'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-Market-Catalog-%E2%80%A2-Product-Page?node-id=1902%3A1900',
        }),
        docs: {
            source: {
                type: 'code',
            },
        },
    }}
/>

export const Template = (args, {argTypes}) => ({
    props: Object.keys(argTypes),
    components: {SPriceRange},
    template: `<s-price-range style="max-width: 200px;" :value="currentValue" @change="changePrice" :max="max" :min="min"/>`,
    data() {
        return {
            currentValue: this.value
        }
    },
    watch: {
        value(v) {
            this.currentValue = v
        }
    },
    methods: {
        changePrice(value) {
            action('change')(value)
            this.currentValue = value
        },
    },
})

<Canvas withToolbar withSource="none">
    <Story name="SPriceRange">
        {Template.bind({})}
    </Story>
</Canvas>


<Description of={SPriceRange} />
<ArgsTable of={SPriceRange} />
<Source code={`<s-price-range :value="[10, 30]"  :max="100" :min="0"/>`} />

