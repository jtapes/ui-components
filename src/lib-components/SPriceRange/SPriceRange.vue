<template>
  <div class="s-price-range">
    <div class="d-f mb-16">
      <s-input
        v-mask="'#*'"
        :value="localValue[0]"
        :max="max"
        :min="min"
        type="number"
        @change="changeInput($event, 0)"
      >
        <template #prepend>
          <span class="ml-16 color-font-secondary">от</span>
        </template>
      </s-input>
      <s-input
        v-mask="'#*'"
        :value="localValue[1]"
        :max="max"
        :min="min"
        type="number"
        class="ml-8"
        @change="changeInput($event, 1)"
      >
        <template #prepend>
          <span class="ml-16 color-font-secondary">до</span>
        </template>
      </s-input>
    </div>
    <s-slider
      v-model="localValue"
      :tooltip="'none'"
      :max="max"
      :min="min"
      :enable-cross="false"
      @drag-start="handleDragStart"
      @drag-end="handleDragEnd"
      @change="handleValueChange"
    />
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Model, Watch } from 'vue-property-decorator'
import SSlider from '../SSlider/SSlider.vue'
import SInput from '../SInput/SInput.vue'

/** Компонент для отображения цены */
@Component({
  components: { SSlider, SInput },
})
export default class SPriceRange extends Vue {
  /** Диапозон цен */
  @Model('change', {
    type: Array,
    validator(array: [number, number]): boolean {
      return array.length === 2
    },
  })
  value!: [number, number]

  /** Минимальная цена */
  @Prop(Number)
  min!: number

  /** Максимальная цена */
  @Prop(Number)
  max!: number

  private localValue: [number, number] | [] = []
  private dragging = false

  @Watch('value')
  onValueChange(value: [number, number]): void {
    if (value) {
      const isDifference = value.some(
        (value, index) => value !== this.localValue[index]
      )

      if (isDifference) {
        this.localValue = value
      }
    } else {
      this.localValue = [this.min, this.max]
    }
  }

  @Watch('min')
  onMinChange(value: number): void {
    this.localValue = [
      value,
      this.localValue[1] ? this.localValue[1] : this.max,
    ]
  }

  @Watch('max')
  onMaxChange(value: number): void {
    this.localValue = [
      this.localValue[0] ? this.localValue[0] : this.min,
      value,
    ]
  }

  changeInput(price: number, index: number): void {
    if (this.localValue.length) {
      if (this.min > price) price = this.min
      if (this.max < price) price = this.max
      if (index === 0) {
        if (this.localValue[1] < price) {
          price = this.localValue[1]
        }
        this.$emit('change', [price, this.localValue[1]])
      }
      if (index === 1) {
        if (this.localValue[0] > price) {
          price = this.localValue[0]
        }
        this.$emit('change', [this.localValue[0], price])
      }
    }
  }

  handleDragStart(): void {
    this.dragging = true
  }

  handleDragEnd(): void {
    this.dragging = false
    this.$emit('change', this.localValue)
  }

  handleValueChange(): void {
    if (!this.dragging) {
      this.$emit('change', this.localValue)
    }
  }

  mounted(): void {
    if (this.value) {
      this.localValue = this.value

      if (this.min > this.localValue[0]) {
        this.$emit('change', [this.min, this.localValue[1]])
      }
      if (this.max < this.localValue[1]) {
        this.$emit('change', [this.localValue[0], this.max])
      }
    } else {
      this.localValue = [this.min, this.max]
    }
  }
}
</script>
