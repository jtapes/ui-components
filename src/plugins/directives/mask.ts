import Vue from 'vue'
import { DirectiveBinding } from 'vue/types/options'

export type TokenType = {
  [key: string]: {
    pattern?: RegExp
    repeat?: boolean
  }
}
export type MaskType = string | { mask: string; tokens: TokenType }

interface MaskedElement extends HTMLInputElement {
  mask?: MaskType
}

const TOKENS = {
  '#': { pattern: /[0-9]/ },
  x: { pattern: /[0-9a-zA-Z]/ },
  X: { pattern: /[0-9а-яА-Яa-zA-Z]/ },
  S: { pattern: /[a-zA-Z]/ },
  E: { pattern: /[0-9a-zA-Z-_@.]/ },
  '*': { repeat: true },
}

function createEvent(name: string) {
  const event = document.createEvent('Event')
  event.initEvent(name, true, true)

  return event
}

function getInputElement(el: HTMLElement) {
  return el instanceof HTMLInputElement ? el : el.querySelector('input')
}

function masking(value: string, mask: string, tokens: TokenType): string {
  let iValue = 0
  let iMask = 0
  let output = ''

  while (iValue < value.length && iMask < mask.length) {
    const token = mask[iMask]
    const maskRegex = tokens[token]
    const currentValue = value[iValue]

    if (maskRegex && maskRegex.pattern) {
      if (maskRegex.pattern.test(currentValue)) {
        output += currentValue
        iMask++
      }

      iValue++
    } else if (maskRegex && maskRegex.repeat) {
      const prevToken = mask[iMask - 1]
      const prevMaskRegex = tokens[prevToken]

      if (prevMaskRegex && prevMaskRegex.pattern) {
        if (!prevMaskRegex.pattern.test(currentValue)) {
          iMask++
        } else {
          iMask--
        }
      }
    } else {
      output += token
      if (currentValue === token) iValue++

      iMask++
    }
  }

  return output
}

function handleValueChange(e: Event) {
  const target = e.target as MaskedElement
  const { mask } = target
  const oldValue = target.value
  if (mask) {
    if (typeof mask === 'string') {
      target.value = masking(oldValue, mask, TOKENS)
    } else {
      target.value = masking(oldValue, mask.mask, mask.tokens)
    }

    if (target.value !== oldValue) {
      target.dispatchEvent(createEvent('input'))
    }
  }
}

function toggleMask(el: HTMLElement, binding: DirectiveBinding) {
  const input = getInputElement(el) as MaskedElement

  if (input && binding.value) {
    input.mask = binding.value
    input.addEventListener('input', handleValueChange)

    if (input.value) {
      input.dispatchEvent(createEvent('input'))
    }
  }
}

export default () => {
  Vue.directive('mask', {
    update(el, binding) {
      if (binding.value && binding.value !== binding.oldValue) {
        Vue.nextTick(() => {
          toggleMask(el, binding)
        })
      }
    },

    bind(el, binding) {
      if (binding.value) {
        toggleMask(el, binding)
      }
    },

    unbind(el, binding) {
      if (binding.value) {
        const input = getInputElement(el)
        if (input) {
          input.removeEventListener('input', handleValueChange)
        }
      }
    },
  })
}
