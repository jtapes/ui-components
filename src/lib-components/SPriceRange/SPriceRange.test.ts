import { mount, createLocalVue } from '@vue/test-utils'
import mask from '@/plugins/directives/mask'
import { fireEvent } from '@testing-library/dom'
import SPriceRange from '@/lib-components/SPriceRange/SPriceRange.vue'

describe('SPriceRange', () => {
  const localVue = createLocalVue()
  localVue.use(mask)

  const wrapper = mount(SPriceRange, {
    localVue,
    propsData: {
      value: [10, 30],
      min: 0,
      max: 100,
    },
  })

  test('init', () => {
    const inputs = wrapper.findAll('input')
    const inputStartPrice = inputs.at(0).element as HTMLInputElement
    const inputEndPrice = inputs.at(1).element as HTMLInputElement

    expect(inputStartPrice.value).toEqual('10')
    expect(inputEndPrice.value).toEqual('30')
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Change input', async () => {
    const inputs = wrapper.findAll('input')
    const inputStartPrice = inputs.at(0).element as HTMLInputElement
    const inputEndPrice = inputs.at(1).element as HTMLInputElement

    await fireEvent.input(inputStartPrice, { target: { value: '35' } })
    await fireEvent.input(inputEndPrice, { target: { value: '5' } })
    const emitted = wrapper.emitted('change')
    expect(emitted).toBeTruthy()
  })
})
