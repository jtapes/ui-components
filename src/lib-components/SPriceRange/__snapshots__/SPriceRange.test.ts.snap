// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SPriceRange init 1`] = `
<div
  class="s-price-range"
>
  <div
    class="d-f mb-16"
  >
    <div
      class="s-input s-input--number s-input--rounded"
    >
      <div
        class="s-input__prepend"
      >
        <span
          class="ml-16 color-font-secondary"
        >
          от
        </span>
      </div>
       
      <input
        class="s-input__inner s-input__inner--prepended"
        max="100"
        min="0"
      />
       
      <!---->
       
      <!---->
       
      <!---->
    </div>
     
    <div
      class="s-input ml-8 s-input--number s-input--rounded"
    >
      <div
        class="s-input__prepend"
      >
        <span
          class="ml-16 color-font-secondary"
        >
          до
        </span>
      </div>
       
      <input
        class="s-input__inner s-input__inner--prepended"
        max="100"
        min="0"
      />
       
      <!---->
       
      <!---->
       
      <!---->
    </div>
  </div>
   
  <div
    class="vue-slider vue-slider-ltr"
    max="100"
    min="0"
    style="padding: 7px 0px; width: auto; height: 4px;"
    tooltip="none"
  >
    <div
      class="vue-slider-rail"
    >
      <div
        class="vue-slider-process"
        style="height: 100%; top: 0px; left: 10%; width: 20%; transition-property: width,left; transition-duration: 0.5s;"
      />
      <div
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="10"
        aria-valuetext="10"
        class="vue-slider-dot"
        role="slider"
        style="width: 14px; height: 14px; transform: translate(-50%, -50%); top: 50%; left: 10%; transition: left 0.5s;"
        tabindex="0"
      >
        <div
          class="vue-slider-dot-handle"
        />
      </div>
      <div
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="30"
        aria-valuetext="30"
        class="vue-slider-dot"
        role="slider"
        style="width: 14px; height: 14px; transform: translate(-50%, -50%); top: 50%; left: 30%; transition: left 0.5s;"
        tabindex="0"
      >
        <div
          class="vue-slider-dot-handle"
        />
      </div>
    </div>
  </div>
</div>
`;
